File: back_velora/src/main/java/com/velora/customer/CustomerService.java
package com.velora.customer;

import java.util.List;
import java.util.UUID;

import com.velora.auth.dto.UserProfileResponse;
import com.velora.customer.dto.*;
import com.velora.user.*;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
public class CustomerService {

    private final UserRepository users;
    private final CustomerAddressRepository addresses;

    public CustomerService(
            UserRepository users,
            CustomerAddressRepository addresses
    ) {
        this.users = users;
        this.addresses = addresses;
    }

    @Transactional(readOnly = true)
    public UserProfileResponse profile(UUID userId) {
        return UserProfileResponse.from(requireCustomer(userId));
    }

    @Transactional
    public UserProfileResponse updateProfile(
            UUID userId,
            CustomerProfileUpdateRequest request
    ) {
        UserEntity user = requireCustomer(userId);

        if (
            request.customerType() == CustomerType.B2B &&
            (
                isBlank(request.businessName()) ||
                isBlank(request.taxId())
            )
        ) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST,
                    "Los clientes B2B deben registrar razón social y NIT."
            );
        }

        user.setFirstName(request.firstName());
        user.setLastName(request.lastName());
        user.setPhone(request.phone());
        user.setCustomerType(request.customerType());

        if (request.customerType() == CustomerType.B2B) {
            user.setBusinessName(request.businessName());
            user.setTaxId(request.taxId());
        } else {
            user.setBusinessName(null);
            user.setTaxId(null);
        }

        return UserProfileResponse.from(users.save(user));
    }

    @Transactional(readOnly = true)
    public List<CustomerAddressResponse> listAddresses(UUID userId) {
        requireCustomer(userId);

        return addresses
                .findAllByUserIdAndActiveTrueOrderByDefaultAddressDescCreatedAtAsc(userId)
                .stream()
                .map(CustomerAddressResponse::from)
                .toList();
    }

    @Transactional
    public CustomerAddressResponse createAddress(
            UUID userId,
            CustomerAddressRequest request
    ) {
        UserEntity user = requireCustomer(userId);

        List<CustomerAddressEntity> existing =
                addresses.findAllByUserIdAndActiveTrueOrderByDefaultAddressDescCreatedAtAsc(userId);

        boolean shouldBeDefault =
                existing.isEmpty() || request.defaultAddress();

        if (shouldBeDefault) {
            clearDefault(existing);

            // La restricción de PostgreSQL permite una sola dirección
            // predeterminada activa por cliente. Forzamos primero el UPDATE
            // de la dirección anterior antes de insertar la nueva.
            addresses.flush();
        }

        CustomerAddressEntity address = new CustomerAddressEntity();
        address.setUser(user);
        apply(address, request);
        address.setDefaultAddress(shouldBeDefault);
        address.setActive(true);

        return CustomerAddressResponse.from(addresses.save(address));
    }

    @Transactional
    public CustomerAddressResponse updateAddress(
            UUID userId,
            UUID addressId,
            CustomerAddressRequest request
    ) {
        requireCustomer(userId);

        CustomerAddressEntity address =
                requireAddress(userId, addressId);

        if (request.defaultAddress()) {
            List<CustomerAddressEntity> existing =
                    addresses.findAllByUserIdAndActiveTrueOrderByDefaultAddressDescCreatedAtAsc(userId);

            clearDefault(existing);

            // Evita violar temporalmente el índice único al cambiar
            // cuál es la dirección predeterminada.
            addresses.flush();

            address.setDefaultAddress(true);
        }

        apply(address, request);

        return CustomerAddressResponse.from(addresses.save(address));
    }

    @Transactional
    public void deleteAddress(
            UUID userId,
            UUID addressId
    ) {
        requireCustomer(userId);

        CustomerAddressEntity address =
                requireAddress(userId, addressId);

        boolean wasDefault = address.isDefaultAddress();

        address.setDefaultAddress(false);
        address.setActive(false);
        addresses.save(address);

        if (wasDefault) {
            List<CustomerAddressEntity> remaining =
                    addresses.findAllByUserIdAndActiveTrueOrderByDefaultAddressDescCreatedAtAsc(userId);

            if (!remaining.isEmpty()) {
                CustomerAddressEntity nextDefault = remaining.getFirst();
                nextDefault.setDefaultAddress(true);
                addresses.save(nextDefault);
            }
        }
    }

    private UserEntity requireCustomer(UUID userId) {
        UserEntity user = users.findById(userId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Cliente no encontrado."
                ));

        if (user.getRole() != UserRole.CUSTOMER) {
            throw new ResponseStatusException(
                    HttpStatus.FORBIDDEN,
                    "La operación está disponible únicamente para clientes."
            );
        }

        return user;
    }

    private CustomerAddressEntity requireAddress(
            UUID userId,
            UUID addressId
    ) {
        return addresses
                .findByIdAndUserIdAndActiveTrue(addressId, userId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Dirección no encontrada."
                ));
    }

    private void clearDefault(
            List<CustomerAddressEntity> existing
    ) {
        for (CustomerAddressEntity address : existing) {
            if (address.isDefaultAddress()) {
                address.setDefaultAddress(false);
            }
        }
    }

    private void apply(
            CustomerAddressEntity address,
            CustomerAddressRequest request
    ) {
        address.setLabel(request.label());
        address.setRecipientName(request.recipientName());
        address.setRecipientPhone(request.recipientPhone());
        address.setDepartment(request.department());
        address.setCity(request.city());
        address.setZone(request.zone());
        address.setAddressLine(request.addressLine());
        address.setReference(request.reference());
    }

    private boolean isBlank(String value) {
        return value == null || value.isBlank();
    }
}

File: apps/back/src/customer/customer.service.ts
import { Injectable } from "@nestjs/common";
import type {
  CustomerAddressRequest, CustomerAddressResponse, CustomerFavoriteResponse,
  CustomerProfileUpdateRequest, UserProfile,
} from "@velora/contracts";
import type { AuthPrincipal } from "../auth/security.js";
import { ApiHttpError } from "../common/http/api-http.error.js";
import { CustomerRepository } from "./customer.repository.js";

@Injectable()
export class CustomerService {
  constructor(private readonly customers: CustomerRepository) {}

  async profile(principal: AuthPrincipal): Promise<UserProfile> {
    return this.requireCustomer(principal);
  }

  async updateProfile(principal: AuthPrincipal, request: CustomerProfileUpdateRequest): Promise<UserProfile> {
    await this.requireCustomer(principal);
    const updated = await this.customers.updateProfile(principal.userId, request);
    if (!updated) throw new ApiHttpError(404, "Cliente no encontrado.");
    return updated;
  }

  async addresses(principal: AuthPrincipal): Promise<CustomerAddressResponse[]> {
    await this.requireCustomer(principal);
    return this.customers.listAddresses(principal.userId);
  }

  async createAddress(principal: AuthPrincipal, request: CustomerAddressRequest): Promise<CustomerAddressResponse> {
    await this.requireCustomer(principal);
    return this.customers.createAddress(principal.userId, request);
  }

  async updateAddress(principal: AuthPrincipal, addressId: string, request: CustomerAddressRequest): Promise<CustomerAddressResponse> {
    await this.requireCustomer(principal);
    const updated = await this.customers.updateAddress(principal.userId, addressId, request);
    if (!updated) throw new ApiHttpError(404, "Dirección no encontrada.");
    return updated;
  }

  async deleteAddress(principal: AuthPrincipal, addressId: string): Promise<void> {
    await this.requireCustomer(principal);
    if (!(await this.customers.deleteAddress(principal.userId, addressId))) {
      throw new ApiHttpError(404, "Dirección no encontrada.");
    }
  }

  async favorites(principal: AuthPrincipal): Promise<CustomerFavoriteResponse[]> {
    await this.requireCustomer(principal);
    return this.customers.listFavorites(principal.userId);
  }

  async addFavorite(principal: AuthPrincipal, productId: string): Promise<CustomerFavoriteResponse> {
    await this.requireCustomer(principal);
    const result = await this.customers.addFavorite(principal.userId, productId);
    if (result === "PRODUCT_NOT_FOUND") throw new ApiHttpError(404, "Producto no encontrado o inactivo.");
    return result;
  }

  async removeFavorite(principal: AuthPrincipal, productId: string): Promise<void> {
    await this.requireCustomer(principal);
    await this.customers.removeFavorite(principal.userId, productId);
  }

  private async requireCustomer(principal: AuthPrincipal): Promise<UserProfile> {
    if (principal.role !== "CUSTOMER") throw new ApiHttpError(403, "La operación está disponible únicamente para clientes.");
    const profile = await this.customers.profile(principal.userId);
    if (!profile) throw new ApiHttpError(401, "Cliente autenticado no encontrado.");
    if (profile.role !== "CUSTOMER") throw new ApiHttpError(403, "La operación está disponible únicamente para clientes.");
    if (profile.status !== "ACTIVE") throw new ApiHttpError(403, "La cuenta del cliente no está activa.");
    return profile;
  }
}
